### PR TITLE
package: add kmod-led-group-multicolor

### PR DESCRIPTION
**Pull request by professor-jonny**
These drivers add suport for multicolour led's and support for combining several monochromatic LEDs into one multi-color
  LED using the multicolor LED class.
This driver is useful in the case where LEDs of
 different colors are physically grouped in a single multi-color LED

Example of use from shell:
```
echo 255 0 0 > /sys/class/leds/rgb:status/multi_intensity  (green)
echo 0 255 0 > /sys/class/leds/rgb:status/multi_intensity  (red)
echo 0 0 255 > /sys/class/leds/rgb:status/multi_intensity (blue)

```

you may simply append a line to /etc/rc.local the below will make the rgb:status led green:

`[[ -f /sys/class/leds/rgb:status/multi_intensity ]] && echo '255 0 0' > /sys/class/leds/rgb:status/multi_intensity`

Example config of dts:
 ```
	monochromatic-leds {
		compatible = "gpio-leds";

		led_system_green: green {
			gpios = <&tlmm 50 GPIO_ACTIVE_HIGH>; 
			color = <LED_COLOR_ID_GREEN>;
			function = LED_FUNCTION_STATUS;
		};

		led_system_red: red {
			gpios = <&tlmm 51 GPIO_ACTIVE_HIGH>;
			color = <LED_COLOR_ID_RED>;
			function = LED_FUNCTION_STATUS;
		};

		led_system_blue: blue {
			gpios = <&tlmm 52 GPIO_ACTIVE_HIGH>;
			color = <LED_COLOR_ID_BLUE>;
			function = LED_FUNCTION_STATUS;
		};
	};

	multicolor-led {
    compatible = "leds-group-multicolor";
		color = <LED_COLOR_ID_RGB>;
		function = LED_FUNCTION_STATUS;
		max-brightness = <256>;
		leds = <&led_system_green>,<&led_system_red>,<&led_system_blue>;
	};
```

Signed-off-by: Jonathan Brophy professor_jonny@hotmail.com

### CODE REVIEW

**Comment by robimarko** [comment]
No, instead of copy/pasting the upstream driver source into OpenWrt, simply backport the upstream driver for 6.1 since 6.6 already has it and just package the required kmod.

**Comment by testuser7** [comment]
Is it possible to use multicolor configuration for reporting status in dts (`led-running`, `led-failsafe`, `led-upgrade`)?

**Comment by professor-jonny** [comment]
Hi @robimarko I'm not sure how to do that I looked for an example and had a look on the wiki.

**Comment by professor-jonny** [comment]
@testuser7 no sorry this just groups single color LED's into a single multicolor class like a multicolour LED controller.

I'm guessing you want to be able to drive a specific color directly from the DTS which is not possible a driver would have to be made to support virtual LED's from the DTS as a mix of individual LED elements.

**Comment by testuser7** [comment]
@professor-jonny There is an option to enable this in kernel config: https://github.com/openwrt/openwrt/blob/main/target/linux/generic/config-6.6#L3044

**Comment by professor-jonny** [comment]
@testuser7 ok so I don't need this and I can enable it in kernel config?
do I still need to patch make files or something I don't really know how it all works.

**Comment by leobsky** [comment]
@professor-jonny patches are only needed for older kernel than 6.6 , perhaps skipping older kernel there is no need of backports the patches and the LED kernel module can be added in` package/kernel/linux/modules/leds.mk`
https://github.com/openwrt/openwrt/blob/main/package/kernel/linux/modules/leds.mk
```
define KernelPackage/led-group-multi-color
  SUBMENU:=$(LEDS_MENU)
  TITLE:=LEDs group multi-color support
  KCONFIG:= \
      CONFIG_LEDS_CLASS_MULTICOLOR=y \
      CONFIG_LEDS_GROUP_MULTICOLOR
  DEPENDS:=+LINUX_6_6
  FILES:=$(LED_TRIGGER_DIR)/rgb/leds-group-multicolor.ko
  AUTOLOAD:=$(call AutoLoad,50,led-group-multi-color)
endef

define KernelPackage/led-group-multi-color/description
 This option enables support for monochrome LEDs that are grouped
 into multicolor LEDs which is useful in the case where LEDs of
 different colors are physically grouped in a single multi-color LED
 and driven by a controller that does not have multi-color support.
endef

$(eval $(call KernelPackage,led-group-multi-color))
```
just an idea, developers can give you better help.

**Comment by robimarko** [comment]
Yes, go into that direction, ignore 5.15 and 6.1 for now, just make the kmod depend on 6.6 as we are going to drop both 5.15 and 6.1 rather soon.

Some quick stuff I see to change:
1. `DEPENDS:=@LINUX_6_6`
2. `AUTOLOAD:=$(call AutoProbe,led-group-multi-color)`
3. `LEDS_CLASS_MULTICOLOR` should be split into another kmod as its also tristate so there is no need to have it built in and you can just add it as a dependency to this one

**Comment by professor-jonny** [comment]
leds-class-multicolor is already enabled in kernel config it is a dependency on many of the i2c and spi rgb led controllers.

**Comment by professor-jonny** [comment]
Is this ok? now
Also I looked in the wiki and it says one package per commit does that apply in this case or can I aslo add led_class_multicolor to this commit?

File: package/kernel/linux/modules/leds.mk [hunk 1]
```diff
@@ -16,6 +16,27 @@ define KernelPackage/leds-gpio
...
 endef
 
+
+define KernelPackage/led-group-multicolor
+	 SUBMENU:=$(LEDS_MENU)
+	 TITLE:=LEDs group multi-color support
+	 KCONFIG:= \
+	     CONFIG_LEDS_CLASS_MULTICOLOR=y \
+	     CONFIG_LEDS_GROUP_MULTICOLOR
+	 DEPENDS:=@LINUX_6_6
```
Comment (hauke) [inline package/kernel/linux/modules/leds.mk:26]: OpenWrt only supports kernel 6.6, please remove this line.

File: package/kernel/linux/modules/leds.mk [hunk 2]
```diff
@@ -16,6 +16,27 @@ define KernelPackage/leds-gpio
...
+endef
+
+define KernelPackage/led-group-multi-color/description
+	 This option enables support for monochrome LEDs that are grouped
+	 into multicolor LEDs which is useful in the case where LEDs of
+	 different colors are physically grouped in a single multi-color LED
+	 and driven by a controller that does not have multi-color support.
+endef
+
+	$(eval $(call KernelPackage,led-group-multi-color))
```
Comment (hauke) [inline package/kernel/linux/modules/leds.mk:38]: Fix indention 
```suggestion
$(eval $(call KernelPackage,led-group-multi-color))
```

File: package/kernel/linux/modules/leds.mk [hunk 3]
```diff
@@ -16,6 +16,27 @@ define KernelPackage/leds-gpio
   AUTOLOAD:=$(call AutoLoad,60,leds-gpio,1)
 endef
 
+
+define KernelPackage/led-group-multicolor
+	 SUBMENU:=$(LEDS_MENU)
+	 TITLE:=LEDs group multi-color support
+	 KCONFIG:= \
+	     CONFIG_LEDS_CLASS_MULTICOLOR=y \
```
Comment (hauke) [inline package/kernel/linux/modules/leds.mk:24]: Please build this as a module too:
```suggestion
	     CONFIG_LEDS_CLASS_MULTICOLOR \
```

File: package/kernel/linux/modules/leds.mk [hunk 4]
```diff
@@ -16,6 +16,27 @@ define KernelPackage/leds-gpio
...
+	     CONFIG_LEDS_GROUP_MULTICOLOR
+	 DEPENDS:=@LINUX_6_6
+	 FILES:=$(LED_TRIGGER_DIR)/rgb/leds-group-multicolor.ko
```
Comment (hauke) [inline package/kernel/linux/modules/leds.mk:27]: I do not think this is working.
I think you need this:
```suggestion
	 FILES:=$(LINUX_DIR)/drivers/leds/rgb/leds-group-multicolor.ko
```

Comment (professor-jonny) [reply package/kernel/linux/modules/leds.mk]: I have done the changes but I will have to test build when I get home

File: package/kernel/linux/modules/leds.mk [hunk 5]
```diff
@@ -16,6 +36,26 @@ define KernelPackage/leds-gpio
   AUTOLOAD:=$(call AutoLoad,60,leds-gpio,1)
 endef
 
+
+define KernelPackage/led-group-multicolor
+	 SUBMENU:=$(LEDS_MENU)
+	 TITLE:=LEDs group multi-color support
+	 KCONFIG:= \
+	     CONFIG_LEDS_CLASS_MULTICOLOR \
```
Comment (robimarko) [inline package/kernel/linux/modules/leds.mk:44]: No, you have a separate kmod for this, just include it as a dependency.

Comment (professor-jonny) [reply package/kernel/linux/modules/leds.mk:29]: Ok I have removed that, and it test builds fine.

**Comment by blogic** [comment]
@professor-jonny do you plan to push with the fixes @hauke requested ?

**Comment by blogic** [comment]
merged, Thanks !

**Comment by professor-jonny** [comment]
> @professor-jonny do you plan to push with the fixes @hauke requested ?

I have done it now sorry I forgot about this, I have been prioritizing developing and testing my own custom driver to overcome the issues with this driver as has limited uses in OpenWrt as multicolor class is not directly supported in Luci and requires scripting or as such to make use of it.

I have created a new led driver that creates virtual single led's by grouping physical monochromatic led's, it also has some things to make it useful as a status led such as priority, logic and on delay/ off delay that can be controlled from sysfs or set in the dts

This also means this it works with aliases and the standard diag.sh triggers and you can define and control led's from Luci with netdev for example.

**Pull request closed without merging**